cdap-app-fabric/src/test/java/co/cask/cdap/WorkflowApp.java

```diff
@@ -120,20 +124,44 @@ public void run(SparkContext context) {
       List<Integer> data = Arrays.asList(1, 2, 3, 4, 5);
       JavaRDD<Integer> distData = ((JavaSparkContext) context.getOriginalSparkContext()).parallelize(data);
       distData.collect();
+
       // If there are problems accessing workflow token here, Spark will throw a NotSerializableException in the test
       final WorkflowToken workflowToken = context.getWorkflowToken();
       if (workflowToken != null) {
-        workflowToken.put("otherKey", "otherValue");
+        workflowToken.put("multiplier", "2");
       }
-      distData.map(new Function<Integer, Integer>() {
+
+      final boolean addToToken = context.getRuntimeArguments().containsKey("addToTokenInFunction");
+
+      JavaRDD<Integer> mapData = distData.map(new Function<Integer, Integer>() {
         @Override
         public Integer call(Integer val) throws Exception {
-          if (workflowToken != null && workflowToken.get("tokenKey") != null) {
-            return 2 * val;
+          if (addToToken) {
+            workflowToken.put("some.key", "some.value");
           }
-          return val;
+
+          if (workflowToken.get("multiplier") != null) {
+            int multiplier = workflowToken.get("multiplier").getAsInt();
+            return multiplier * val;
+          }
+          return 0;
         }
       });
+
+      boolean exceptionThrown = false;
+      try {
+        List<Integer> result = mapData.collect();
+        Assert.assertTrue(result.equals(Arrays.asList(2, 4, 6, 8, 10)));
+      } catch (Throwable t) {
+        exceptionThrown = true;
+      }
+
+      if (addToToken) {
+        Assert.assertTrue(exceptionThrown);
```

Comment:
If the closure throws exception, the spark program execution should be in failed state. It's better to verify that from the test method, otherwise it's not really verifying.